src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.InlineContext.cs

```diff
@@ -0,0 +1,582 @@
+                // {
+                //     DoSomething(out j, out int i);
+                // }
+                // After:
+                // void Caller()
+                // {
+                //     int i = 10;
+                //     DoSomething(out i, out int i2);
+                // }
+                // void Callee(out int j)
+                // {
+                //     DoSomething(out j, out int i);
+                // }
+                var localDeclarationStatementsNeedInsert = GetLocalDeclarationStatementsNeedInsert(
+                    inlineMethodRefactoringProvider,
+                    syntaxGenerator,
+                    methodParametersInfo.OperationsNeedGenerateDeclarations,
+                    parametersWithVariableDeclarationArgumentToName!,
+                    renameTable);
+
+                // Get a table which the key is the symbol needs replacement work. Value is the replacement syntax node
+                var replacementTable = ComputeReplacementTable(
+                    inlineMethodRefactoringProvider,
+                    calleeMethodSymbol,
+                    methodParametersInfo.ParametersWithLiteralArgument,
+                    methodParametersInfo.ParametersWithDefaultValue,
+                    syntaxGenerator,
+                    renameTable);
+
+                // Do the replacement work within the callee's body so that it can be inserted to the caller later.
+                inlineNode = (TExpression)await ReplaceAllSyntaxNodesForSymbolAsync(
+                   document,
+                   inlineNode,
+                   syntaxGenerator,
+                   root,
+                   replacementTable,
+                   cancellationToken).ConfigureAwait(false);
+
+                // Check if there is await expression. It is used later if the caller should be changed to async
+                var containsAwaitExpression = inlineNode
+                    .DescendantNodesAndSelf()
+                    .Any(node => node != null && syntaxFacts.IsAwaitExpression(node));
+
+                if (syntaxFacts.IsExpressionStatement(statementContainsCallee)
+                    && !calleeMethodSymbol.ReturnsVoid
+                    && !syntaxFacts.IsArgument(calleeInvocationNode.Parent))
+                {
+                    // If the callee is invoked like
+                    // void Caller()
+                    // {
+                    //     Callee();
+                    // }
+                    // int Callee()
+                    // {
+                    //     return 1;
+                    // };
+                    // After it should be:
+                    // void Caller()
+                    // {
+                    //     int temp = 1;
+                    // }
+                    // int Callee()
+                    // {
+                    //     return 1;
+                    // };
+                    // One variable declaration needs to be generated.
```

Comment:
i don't know if i agree with this.  It's not really if the callee retursn void or something.  for example if i have:

```
void Caller(HashSet<int> h)
{
    Callee(h);
}
bool Callee()
{
    return h.Add(0);
}
```

This shuldn't become:

```
void Caller(HashSet<int> h)
{
    int temp = h.Add(0);
}
```

the rule here is not about void/non-void.  It's if the expression itself is a syntactic form that c# diallows as an expressoin-statement.  i.e. `h.Add(0);` is fine in c#.  but `0;` is not.

Reply:
I see, so from the example you show then `return h.Add(0);` is an invocation statement within a return statement.
After that if it is put seperately `void Caller() { h.Add(0); }` it is an ExpressionStatement (which is allowed) 
Then the correct check would be, if after the replacement the statement will becomes an ExpressionStatement.  If it is not and it's return type is not void then a temp variable should be generated.

Reply:
From your example `h.add(0)` is invocation expressoin. And `Callee(h)` is invocation expression. So after replacement it becomes another ExpressionStatement which is allowed in C#. (Looks like VB has the same syntax node [here ](http://sourceroslyn.io/#Microsoft.CodeAnalysis.VisualBasic/Generated/Syntax.xml.Syntax.Generated.vb,e5c9d2df19f9ee15)but only allow await and invocation)
In this [section](https://github.com/dotnet/csharplang/blob/master/spec/statements.md#expression-statements) from C# lang
ExpressionStatements is
expression_statement
    : statement_expression ';'
    ;

statement_expression
    : invocation_expression
    | null_conditional_invocation_expression
    | object_creation_expression
    | assignment
    | post_increment_expression
    | post_decrement_expression
    | pre_increment_expression
    | pre_decrement_expression
    | await_expression
    ;

Would it be ok to check if the inline content syntax node's kind, if it is one of there expression then no 'temp' is required? (Similarly check for VB)
It will require us to sync the compiler side change (in case they add more valid expression)

Reply:
Yes. I think that would be ok.